docs/abi.rst

```diff
@@ -0,0 +1,560 @@
+        return For(
+            i.store(Int(0)), i.load() < array.length(), i.store(i.load() + Int(1))
+        ).Do(
+            array[i.load()].use(lambda value: Assert(value.get() > Int(5)))
+        )
+
+Limitations
+"""""""""""""""""""""
+
+TODO: explain type size limitations
+
+Reference Types
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+Some AVM operations require specific values to be placed in the "foreign arrays" of the app call transaction. Reference types allow methods to describe these requirements.
+
+Reference types are only valid in the arguments of a method. They may not appear in a method's return value in any form.
+
+Definitions
+""""""""""""""""""""""""""""""""""""""""""
+
+PyTeal supports the following reference types:
+
+====================== ====================== ================ =======================================================================================================================================================
+PyTeal Type            ARC-4 Type             Dynamic / Static Description
+====================== ====================== ================ =======================================================================================================================================================
+:any:`abi.Account`     :code:`account`        Static           Represents an additional account that the current transaction can access, stored in the :any:`Txn.accounts <TxnObject.accounts>` array
+:any:`abi.Asset`       :code:`asset`          Static           Represents an additional asset that the current transaction can access, stored in the :any:`Txn.assets <TxnObject.assets>` array
+:any:`abi.Application` :code:`application`    Static           Represents an additional application that the current transaction can access, stored in the :any:`Txn.applications <TxnObject.applications>` array
+====================== ====================== ================ =======================================================================================================================================================
+
+These types all inherit from the abstract class :any:`abi.ReferenceType`.
+
+Usage
+""""""""""""""""""""""""""""""""""""""""""
+
+Getting Referenced Indexes
+''''''''''''''''''''''''''
+
+Because reference types represent values placed into one of the transaction's foreign arrays, each reference type value is associated with a specific index into the appropriate array.
+
+All reference types implement the method :any:`abi.ReferenceType.referenced_index()` which can be used to access this index.
+
+A brief example is below:
+
+.. code-block:: python
+
+    @Subroutine(TealType.none)
+    def referenced_index_example(
+        account: abi.Account, asset: abi.Asset, app: abi.Application
+    ) -> Expr:
+        return Seq(
+            # The accounts array has Txn.accounts.length() + 1 elements in it (the +1 is the txn sender)
+            Assert(account.referenced_index() <= Txn.accounts.length()),
```

Comment:
it took me a few minutes to understand how `referenced_index()` works, still a bit confused maybe about what the point of it is

Reply:
ok, so this example method requires an account reference, an asset reference, and an app reference as arguments. 

These aren't all at index 0, though, because for accounts and apps there's a default at index 0. 

The comment "The accounts array has Txn.accounts.length() + 1 elements in it" - you're saying the length() method is implemented to actually not give what you'd expect, which is the length of that array, but rather the length of the args put in that array? 

In short: I think you should add a little more explanation and comments here because it feels pretty counterintuitive.  And maybe the example could have 2 args of one of the arrays to help understand.